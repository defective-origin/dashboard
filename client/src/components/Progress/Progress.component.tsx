import React from 'react'
import MuiCircularProgress from '@mui/material/CircularProgress'
import MuiLinearProgress from '@mui/material/LinearProgress'

// ---| core |---
import { cn, react } from 'tools'

// ---| self |---
import css from './Progress.module.scss'

export const PROGRESS_MAP = {
  circular: MuiCircularProgress,
  linear: MuiLinearProgress,
}

export type ProgressVariant = keyof typeof PROGRESS_MAP
// todo: add color
export type ProgressProps = {
  className?: string
  show?: boolean
  v?: ProgressVariant
  value?: number
}

/**
 * Component description.
 *
 * How to use
 * @example
 * <Progress />
 */
export function Progress(props: ProgressProps): JSX.Element | null {
  const { show, v = 'circular', value, className, ...otherProps } = props
  const _className = cn(css.Progress, className)
  const Tag = PROGRESS_MAP[v]

  if (!show) {
    return null
  }

  return <Tag className={_className} value={value} {...otherProps} />
}

Progress.displayName = 'Progress'

export default react.attachOverrides(Progress, {
  Circular: { v: 'circular' },
  Linear: { v: 'linear' },
}, {
  memoize: true,
})
