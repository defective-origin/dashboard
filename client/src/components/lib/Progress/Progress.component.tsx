import React from 'react'
import MuiCircularProgress, { CircularProgressProps as MuiCircularProgressProps } from '@mui/material/CircularProgress'
import MuiLinearProgress, { LinearProgressProps as MuiLinearProgressProps } from '@mui/material/LinearProgress'

// ---| common |---
import { cn, UI } from 'common/tools'

// ---| self |---
import css from './Progress.module.scss'

export const PROGRESS_MAP = {
  circular: MuiCircularProgress,
  linear: MuiLinearProgress,
}

export type ProgressType = keyof typeof PROGRESS_MAP
// todo: fix props types
export type ProgressProps = {
  className?: string
  show?: boolean
  type?: ProgressType
}

/**
 * Component description.
 *
 * How to use
 * @example
 * <Progress />
 */
export function Progress(props: ProgressProps): JSX.Element | null {
  const { show, type = 'circular', className, ...otherProps } = props
  const _className = cn(css.Progress, className)
  const Tag = PROGRESS_MAP[type]

  if (!show) {
    return null
  }

  return <Tag className={_className} {...otherProps} />
}

Progress.displayName = 'Progress'

export default UI.attachOverrides(Progress, {
  Circular: { type: 'circular' },
  Linear: { type: 'linear' },
}, {
  memoize: true,
})
