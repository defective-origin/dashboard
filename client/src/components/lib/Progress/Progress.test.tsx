// ---| tests |---
import { render } from 'tests'

// ---| self |---
import Progress from './Progress.component'

describe('[Progress] component', () => {
  it('should render component', () => {
    const container = render(<Progress />)

    expect(container.snapshot()).toMatchSnapshot()
  })
})
