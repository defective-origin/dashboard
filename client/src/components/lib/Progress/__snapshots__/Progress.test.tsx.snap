// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[Progress] component > should render component 1`] = `<DocumentFragment />`;
